samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImplGraph.java

```diff
@@ -38,127 +43,99 @@
 
 
 /**
- * Instantiates the DAG of {@link OperatorImpl}s corresponding to the {@link OperatorSpec}s for the input
- * {@link MessageStreamImpl}s.
+ * The DAG of {@link OperatorImpl}s corresponding to the DAG of {@link OperatorSpec}s.
  */
 public class OperatorImplGraph {
 
   /**
    * A mapping from {@link OperatorSpec}s to their {@link OperatorImpl}s in this graph. Used to avoid creating
-   * multiple {@link OperatorImpl}s for an {@link OperatorSpec}, e.g., when it's reached from different
-   * input {@link MessageStreamImpl}s.
+   * multiple {@link OperatorImpl}s for an {@link OperatorSpec} when it's reached from different
+   * {@link OperatorSpec}s during DAG traversals (e.g., for the merge operator).
    */
   private final Map<OperatorSpec, OperatorImpl> operatorImpls = new HashMap<>();
 
   /**
-   * A mapping from input {@link SystemStream}s to their {@link OperatorImpl} sub-DAG in this graph.
+   * A mapping from input {@link SystemStream}s to their {@link InputOperatorImpl} sub-DAG in this graph.
    */
-  private final Map<SystemStream, RootOperatorImpl> rootOperators = new HashMap<>();
-
-  private final Clock clock;
+  private final Map<SystemStream, InputOperatorImpl> inputOperators = new HashMap<>();
 
-  public OperatorImplGraph(Clock clock) {
-    this.clock = clock;
-  }
+  /**
+   * A mapping from {@link JoinOperatorSpec}s to their two {@link PartialJoinFunction}s. Used to associate
+   * the two {@link PartialJoinOperatorImpl}s for a {@link JoinOperatorSpec} with each other since they're
+   * reached from different {@link OperatorSpec} during DAG traversals.
+   */
+  private final Map<Integer, Pair<PartialJoinFunction, PartialJoinFunction>> joinFunctions = new HashMap<>();
```

Comment:
It's required for correctness. `computeIfAbsent` returns current value if present. We create both partial join implementations the first time we encounter the join operator spec. The next time we see this we get and use the other side.